fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java

```diff
@@ -65,57 +72,32 @@ public OptExpression optimize(ConnectContext connectContext,
-                memo.getRootGroup(), RuleSetType.SUBQUERY_REWRITE));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
 
+        ruleRewriteIterative(memo, rootTaskContext, RuleSetType.MULTI_DISTINCT_REWRITE);
+        ruleRewriteIterative(memo, rootTaskContext, RuleSetType.SUBQUERY_REWRITE);
         // Note: PUSH_DOWN_PREDICATE tasks should be executed before MERGE_LIMIT tasks
         // because of the Filter node needs to be merged first to avoid the Limit node
         // cannot merge
-        context.getTaskScheduler().pushTask(new TopDownRewriteTask(rootTaskContext,
-                memo.getRootGroup(), RuleSetType.PUSH_DOWN_PREDICATE));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
-
-        context.getTaskScheduler().pushTask(new TopDownRewriteTask(rootTaskContext,
-                memo.getRootGroup(), RuleSetType.PRUNE_COLUMNS));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
-
+        ruleRewriteIterative(memo, rootTaskContext, RuleSetType.PUSH_DOWN_PREDICATE);
+        ruleRewriteOnlyOnce(memo, rootTaskContext, RuleSetType.PRUNE_COLUMNS);
+        ruleRewriteIterative(memo, rootTaskContext, new PruneWindowRule());
+        ruleRewriteIterative(memo, rootTaskContext, new MergeTwoProjectRule());
         //Limit push must be after the column prune,
         //otherwise the Node containing limit may be prune
-        context.getTaskScheduler().pushTask(new TopDownRewriteTask(rootTaskContext,
-                memo.getRootGroup(), RuleSetType.MERGE_LIMIT));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
-
-        context.getTaskScheduler().pushTask(new TopDownRewriteTask(rootTaskContext,
-                memo.getRootGroup(), RuleSetType.MERGE_AGGREGATE));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
-
+        ruleRewriteIterative(memo, rootTaskContext, RuleSetType.MERGE_LIMIT);
+        ruleRewriteIterative(memo, rootTaskContext, new MergeTwoAggRule());
         //After the MERGE_LIMIT, ProjectNode that can be merged may appear.
         //So we do another column cropping
-        rootTaskContext.setRequiredColumns((ColumnRefSet) requiredColumns.clone());
-        context.getTaskScheduler().pushTask(new TopDownRewriteTask(rootTaskContext,
-                memo.getRootGroup(), RuleSetType.PRUNE_COLUMNS));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
-
-        context.getTaskScheduler().pushTask(new TopDownRewriteTask(rootTaskContext,
-                memo.getRootGroup(), RuleSetType.PRUNE_ASSERT_ROW));
-        context.getTaskScheduler().executeTasks(rootTaskContext, memo.getRootGroup());
+        ruleRewriteIterative(memo, rootTaskContext, new MergeTwoProjectRule());
+        ruleRewriteIterative(memo, rootTaskContext, RuleSetType.PRUNE_ASSERT_ROW);
```

Comment:
Maybe we can use a MEREG_OPERATOR group to describe these rules?

Reply:
Merge and other rules are in order and cannot be executed in a group